src/Fixer/Import/FullyQualifiedStrictTypesFixer.php

```diff
@@ -262,16 +262,16 @@ private function fixPhpDoc(Tokens $tokens, int $index, array $uses, string $name
     {
         $phpDoc = $tokens[$index];
         $phpDocContent = $phpDoc->getContent();
-        Preg::matchAll('#@([^\s]+)\s+([^\s]+)#', $phpDocContent, $matches);
+        Preg::matchAll('#@([^\s]+)(\s+)((?:[a-zA-Z0-9_\\\\]+)+)#', $phpDocContent, $matches);
```

Comment:
do we need 2 groups?

```diff
-((?:[a-zA-Z0-9_\\\\]+)+)
+([a-zA-Z0-9_\\\\]{2,})
```

```diff
-((?:[a-zA-Z0-9_\\\\]+)+)
+([a-zA-Z0-9_\\\\]+)
```

Reply:
Yes, because it matches any amount of `\Foo` blocks, with single group it only matches the last one.

Reply:
why that big change? why not `'#@([^\s]+)(\s+)([^\s}]+)#'` ?

Reply:
Because the change provided in v3.24 was intended to support only FQCNs in phpDoc tags, but it also matches URLs and other complex types not containing whitespace (see #7629 or try to run this fixer with `import_symbols=true` on our codebase - 2 files are failing). This fix narrows regex to what code actually supports, and wider support can be added later.

Reply:
overall, when changing regexes, I would love to see a testcase that is failing with previous regex but working with new regex.
changing one regex with another without evidence of different work in tests is confusing to me.

here, I'm ok to replace the `[^\s]` with explicit list of allowed chars (but i still wonder what case would fail if we do not do it)

Reply:
> Because the change provided in v3.24 was intended to support only FQCNs in phpDoc tags, but it also matches URLs and other complex types not containing whitespace (see #7629 or try to run this fixer with `import_symbols=true` on our codebase - 2 files are failing). This fix narrows regex to what code actually supports, and wider support can be added later.

@Wirone , I do not understand this part.
I applied simpler regex and everything working as expected. Can you incorporate the new test that show need of this non-capturing subgroup? or remove it


```
ker@d ~/github/FriendsOfPHP/PHP-CS-Fixer λ git log -1
'f215d82e6 2023-12-28 17:40:08 - (HEAD -> codito/fix-fqcn-in-phpdoc) Merge branch 'master' into codito/fix-fqcn-in-phpdoc [Da>
ker@d ~/github/FriendsOfPHP/PHP-CS-Fixer λ git diff src
diff --git a/src/Fixer/Import/FullyQualifiedStrictTypesFixer.php b/src/Fixer/Import/FullyQualifiedStrictTypesFixer.php
index 09bd0eb50..b1822cae1 100644
--- a/src/Fixer/Import/FullyQualifiedStrictTypesFixer.php
+++ b/src/Fixer/Import/FullyQualifiedStrictTypesFixer.php
@@ -262,7 +262,7 @@ private function fixPhpDoc(Tokens $tokens, int $index, array $uses, string $name
     {
         $phpDoc = $tokens[$index];
         $phpDocContent = $phpDoc->getContent();
-        Preg::matchAll('#@([^\s]+)(\s+)((?:[a-zA-Z0-9_\\\\]+)+)#', $phpDocContent, $matches);
+        Preg::matchAll('#@([^\s]+)(\s+)([a-zA-Z0-9_\\\\]+)#', $phpDocContent, $matches);
 
         if ([] !== $matches[0]) {
             foreach ($matches[3] as $i => $typeName) {
ker@d ~/github/FriendsOfPHP/PHP-CS-Fixer λ ./vendor/bin/phpunit tests/Fixer/Import/FullyQualifiedStrictTypesFixerTest.php
PHPUnit 9.6.15 by Sebastian Bergmann and contributors.

.........................................................................S...................                   93 / 93 (100%)

Time: 00:00.316, Memory: 28.50 MB

OK, but incomplete, skipped, or risky tests!
Tests: 93, Assertions: 6585, Skipped: 1.
```

Reply:
```suggestion
        Preg::matchAll('#@([^\s]+)(\s+)([a-zA-Z0-9_\\\\]+)#', $phpDocContent, $matches);
```

Reply:
> overall, when changing regexes, I would love to see a testcase that is failing with previous regex but working with new regex.
changing one regex with another without evidence of different work in tests is confusing to me.

Failing test cases were added in the first commit, CI failed for both of them, further commits provide the fix. I don't understand the confusion here 🤷‍♂️.

Now you merged your change in the regex and this PR needs conflict resolution, which I won't do now because I'm AFK.

Reply:
In terms of regex, I am pretty sure that when I was testing it with regex101 there was a hint about capturing last match vs repeated match, that's why I added it. I can't verify it right now.

Reply:
> Now you merged your change in the regex and this PR needs conflict resolution, which I won't do now because I'm AFK.

before you wrote that, i already synced your PR to not have the conflict.

Reply:
> Failing test cases were added in the first commit, CI failed for both of them, further commits provide the fix. I don't understand the confusion here 🤷‍♂️.

this is about original regex vs new regex.
I'm talking about new regex vs regex I am proposing.


> In terms of regex, I am pretty sure that when I was testing it with regex101 there was a hint about capturing last match vs repeated match, that's why I added it. I can't verify it right now.

I would love to see that by seeing failing test case when using simplified version of regex that I proposed.
I do not see that happening, on my local machine. I would not like to base the code variant on that anecdote. I'm really curious to see it in place, as test case ;)

let's incorporate simplified regex version and if you find out later the case which is failing with simplified version, let's add that test case and recover non-capturing sub-group

Reply:
@keradus FYI: you're right, double capturing group was a leftover from previous regex `@([^\s]+)(\s+)((?:\\[a-zA-Z0-9_]+)+)` which enforced FQCNs with leading backslash (single or repeated `\Foo` blocks). But later I changed it because in global namespace leading backslash is not required, yet I kept more complex regex unnecessarily 🤷‍♂️ .